File: test/language/alignment/python/OptionalMemberAlignmentTest.py
import unittest
import zserio

from testutils import getZserioApi

class OptionalMemberAlignmentTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "alignment.zs").optional_member_alignment

    def testBitSizeOfWithOptional(self):
        optionalMemberAlignment = self.api.OptionalMemberAlignment(True, 0x4433, 0x1122)
        self.assertEqual(self.WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE, optionalMemberAlignment.bitsizeof())

    def testBitSizeOfWithoutOptional(self):
        optionalMemberAlignment = self.api.OptionalMemberAlignment(False, None, 0x7624)
        self.assertEqual(self.WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE, optionalMemberAlignment.bitsizeof())

    def testInitializeOffsetsWithOptional(self):
        optionalMemberAlignment = self.api.OptionalMemberAlignment(True, 0x1111, 0x3333)
        for bitPosition in range(32):
            self.assertEqual(self.WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE,
                             optionalMemberAlignment.initialize_offsets(bitPosition))

        bitPosition = 32
        self.assertEqual(self.WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE + bitPosition,
                         optionalMemberAlignment.initialize_offsets(bitPosition))

    def testInitializeOffsetsWithoutOptional(self):
        optionalMemberAlignment = self.api.OptionalMemberAlignment(field_=0x3334)
        bitPosition = 1
        self.assertEqual(self.WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE + bitPosition,
                         optionalMemberAlignment.initialize_offsets(bitPosition))

    def testReadWithOptional(self):
        hasOptional = True
        optionalField = 0x1234
        field = 0x7654
        writer = zserio.BitStreamWriter()
        OptionalMemberAlignmentTest._writeOptionalMemberAlignmentToStream(writer, hasOptional, optionalField,
                                                                          field)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        optionalMemberAlignment = self.api.OptionalMemberAlignment.from_reader(reader)
        self._checkOptionalMemberAlignment(optionalMemberAlignment, hasOptional, optionalField, field)

    def testReadWithoutOptional(self):
        hasOptional = False
        field = 0x2222
        writer = zserio.BitStreamWriter()
        OptionalMemberAlignmentTest._writeOptionalMemberAlignmentToStream(writer, hasOptional, None, field)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        optionalMemberAlignment = self.api.OptionalMemberAlignment.from_reader(reader)
        self._checkOptionalMemberAlignment(optionalMemberAlignment, hasOptional, None, field)

    def testWriteWithOptional(self):
        hasOptional = True
        optionalField = 0x9ADB
        field = 0x8ACD
        optionalMemberAlignment = self.api.OptionalMemberAlignment(hasOptional, optionalField, field)
        bitBuffer = zserio.serialize(optionalMemberAlignment)
        readOptionalMemberAlignment = zserio.deserialize(self.api.OptionalMemberAlignment, bitBuffer)
        self._checkOptionalMemberAlignment(readOptionalMemberAlignment, hasOptional, optionalField, field)
        self.assertTrue(optionalMemberAlignment == readOptionalMemberAlignment)

    def testWriteWithoutOptional(self):
        hasOptional = False
        field = 0x7ACF
        optionalMemberAlignment = self.api.OptionalMemberAlignment(has_optional_=hasOptional, field_=field)
        bitBuffer = zserio.serialize(optionalMemberAlignment)
        readOptionalMemberAlignment = zserio.deserialize(self.api.OptionalMemberAlignment, bitBuffer)
        self._checkOptionalMemberAlignment(readOptionalMemberAlignment, hasOptional, None, field)
        self.assertTrue(optionalMemberAlignment == readOptionalMemberAlignment)

    @staticmethod
    def _writeOptionalMemberAlignmentToStream(writer, hasOptional, optionalField, field):
        writer.write_bool(hasOptional)
        if hasOptional:
            writer.write_bits(0, 31)
            writer.write_bits(optionalField, 32)
        writer.write_bits(field, 32)

    def _checkOptionalMemberAlignment(self, optionalMemberAlignment, hasOptional, optionalField, field):
        self.assertEqual(hasOptional, optionalMemberAlignment.has_optional)
        if hasOptional:
            self.assertTrue(optionalMemberAlignment.is_optional_field_used())
            self.assertEqual(optionalField, optionalMemberAlignment.optional_field)
        else:
            self.assertFalse(optionalMemberAlignment.is_optional_field_used())
        self.assertEqual(field, optionalMemberAlignment.field)

    WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE = 96
    WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE = 33

File: test/language/alignment/cpp/OptionalMemberAlignmentTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "alignment/optional_member_alignment/OptionalMemberAlignment.h"

namespace alignment
{
namespace optional_member_alignment
{

class OptionalMemberAlignmentTest : public ::testing::Test
{
protected:
    void writeOptionalMemberAlignmentToByteArray(zserio::BitStreamWriter& writer, bool hasOptional,
            int32_t optionalField, int32_t field)
    {
        writer.writeBool(hasOptional);

        if (hasOptional)
        {
            writer.writeBits(0, 31);
            writer.writeSignedBits(optionalField, 32);
        }

        writer.writeSignedBits(field, 32);
    }

    void checkOptionalMemberAlignment(const OptionalMemberAlignment& optionalMemberAlignment, bool hasOptional,
            int32_t optionalField, int32_t field)
    {
        ASSERT_EQ(hasOptional, optionalMemberAlignment.getHasOptional());

        if (hasOptional)
        {
            ASSERT_EQ(optionalField, optionalMemberAlignment.getOptionalField());
            ASSERT_TRUE(optionalMemberAlignment.isOptionalFieldUsed());
        }
        else
        {
            ASSERT_FALSE(optionalMemberAlignment.isOptionalFieldUsed());
        }

        ASSERT_EQ(field, (int)optionalMemberAlignment.getField());
    }

    void fillOptionalMemberAlignment(OptionalMemberAlignment& optionalMemberAlignment, bool hasOptional,
            int32_t optionalField, int32_t field)
    {
        optionalMemberAlignment.setHasOptional(hasOptional);
        if (hasOptional)
            optionalMemberAlignment.setOptionalField(optionalField);
        optionalMemberAlignment.setField(field);
    }

    static const size_t WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    static const size_t WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
};

const size_t OptionalMemberAlignmentTest::WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE = 96;
const size_t OptionalMemberAlignmentTest::WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE = 33;

TEST_F(OptionalMemberAlignmentTest, readWithOptional)
{
    const bool hasOptional = true;
    const int32_t optionalField = 0x1234;
    const int32_t field = 0x7654;
    zserio::BitStreamWriter writer;
    writeOptionalMemberAlignmentToByteArray(writer, hasOptional, optionalField, field);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);

    OptionalMemberAlignment optionalMemberAlignment(reader);
    checkOptionalMemberAlignment(optionalMemberAlignment, hasOptional, optionalField, field);
}

TEST_F(OptionalMemberAlignmentTest, readWithoutOptional)
{
    const bool hasOptional = false;
    const int32_t optionalField = 0;
    const int32_t field = 0x2222;
    zserio::BitStreamWriter writer;
    writeOptionalMemberAlignmentToByteArray(writer, hasOptional, optionalField, field);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);

    OptionalMemberAlignment optionalMemberAlignment(reader);
    checkOptionalMemberAlignment(optionalMemberAlignment, hasOptional, optionalField, field);
}

TEST_F(OptionalMemberAlignmentTest, bitSizeOfWithOptional)
{
    OptionalMemberAlignment optionalMemberAlignment;
    fillOptionalMemberAlignment(optionalMemberAlignment, true, 0x4433, 0x1122);

    const size_t expectedBitSize = WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    ASSERT_EQ(expectedBitSize, optionalMemberAlignment.bitSizeOf());
}

TEST_F(OptionalMemberAlignmentTest, bitSizeOfWithoutOptional)
{
    OptionalMemberAlignment optionalMemberAlignment;
    fillOptionalMemberAlignment(optionalMemberAlignment, false, 0, 0x7624);

    const size_t expectedBitSize = WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    ASSERT_EQ(expectedBitSize, optionalMemberAlignment.bitSizeOf());
}

TEST_F(OptionalMemberAlignmentTest, initializeOffsetsWithOptional)
{
    OptionalMemberAlignment optionalMemberAlignment;
    fillOptionalMemberAlignment(optionalMemberAlignment, true, 0x1111, 0x3333);

    const size_t expectedBitSize = WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    size_t bitPosition = 0;
    for (; bitPosition < 32; ++bitPosition)
        ASSERT_EQ(expectedBitSize, optionalMemberAlignment.initializeOffsets(bitPosition));
    ASSERT_EQ(expectedBitSize + bitPosition, optionalMemberAlignment.initializeOffsets(bitPosition));
}

TEST_F(OptionalMemberAlignmentTest, initializeOffsetsWithoutOptional)
{
    OptionalMemberAlignment optionalMemberAlignment;
    fillOptionalMemberAlignment(optionalMemberAlignment, false, 0, 0x3334);

    const size_t expectedBitSize = WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    const size_t bitPosition = 1;
    ASSERT_EQ(expectedBitSize + bitPosition, optionalMemberAlignment.initializeOffsets(bitPosition));
}

TEST_F(OptionalMemberAlignmentTest, writeWithOptional)
{
    const bool hasOptional = true;
    const int32_t optionalField = 0x9ADB;
    const int32_t field = 0x8ACD;
    OptionalMemberAlignment optionalMemberAlignment;
    fillOptionalMemberAlignment(optionalMemberAlignment, hasOptional, optionalField, field);

    zserio::BitStreamWriter writer;
    optionalMemberAlignment.write(writer);

    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    OptionalMemberAlignment readOptionalMemberAlignment(reader);
    checkOptionalMemberAlignment(readOptionalMemberAlignment, hasOptional, optionalField, field);
    ASSERT_TRUE(optionalMemberAlignment == readOptionalMemberAlignment);
}

TEST_F(OptionalMemberAlignmentTest, writeWithoutOptional)
{
    const bool hasOptional = true;
    const int32_t optionalField = 0;
    const int32_t field = 0x7ACF;
    OptionalMemberAlignment optionalMemberAlignment;
    fillOptionalMemberAlignment(optionalMemberAlignment, hasOptional, optionalField, field);

    zserio::BitStreamWriter writer;
    optionalMemberAlignment.write(writer);

    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    OptionalMemberAlignment readOptionalMemberAlignment(reader);
    checkOptionalMemberAlignment(readOptionalMemberAlignment, hasOptional, optionalField, field);
    ASSERT_TRUE(optionalMemberAlignment == readOptionalMemberAlignment);
}

} // namespace optional_member_alignment
} // namespace alignment
